### PR TITLE
fix: Fix false positives for "unnecessary else" diagnostic

### LINKED ISSUE #16556: `remove-unnecessary-else` bad suggestions and false positives on thread_local macro
**rust-analyzer version**: rust-analyzer 0.3.1839-standalone

**rustc version**: rustc 1.76.0 (07dca489a 2024-02-04)

**relevant settings**: Unlikely any are relevant

Bad suggestions: Given this code

```rust
fn main() {
    let files = [1];

    let query = if files.is_empty() {
        return;
    } else if files.len() == 1 {
        1
    } else {
        17
    };
    print!("{query}");
}
```
remove-unnecessary-else suggests

```rust
fn main() {
    let files = [1];

    let query = if files.is_empty() {
        return;
    }
    if files.len() == 1 {
        1
    } else {
        17
    };
    print!("{query}");
}
```
which doesn't compile.

I also get false positives on `thread_local!{}` macros, but I don't have a minimal reproduction. In my projects it lints on every `thread_local!{}` call, even those as simple as `thread_local!(static SOMETHING: Cell<u32> = Cell::default());` but I don't get the lints in minimal projects.

### PR DESCRIPTION
**Pull request by davidsemakula**
Completes https://github.com/rust-lang/rust-analyzer/pull/16567 by @ShoyuVanilla (see https://github.com/rust-lang/rust-analyzer/pull/16567#discussion_r1492700667)
Fixes #16556

### CODE REVIEW

File: crates/ide-diagnostics/src/handlers/remove_unnecessary_else.rs [hunk 1]
```diff
@@ -387,4 +465,39 @@ fn test() {
...
+"#,
+        );
+        check_diagnostics_with_disabled(
+            r#"
+fn test3() {
+    foo(if a { return 1 } else { 0 })
+}
+"#,
+            std::iter::once("E0308".to_owned()),
+        );
```
Comment (davidsemakula) [inline crates/ide-diagnostics/src/handlers/remove_unnecessary_else.rs:501]: @Veykril is there a better test for this case?

File: crates/hir-ty/src/diagnostics/expr.rs [hunk 2]
```diff
@@ -301,6 +308,36 @@ impl ExprValidator {
...
+                        let ast::Expr::IfExpr(if_expr) = source_ptr.value.to_node(&root) else {
+                            return;
+                        };
+                        let mut top_if_expr = if_expr;
+                        loop {
+                            let parent = top_if_expr.syntax().parent();
+                            let has_parent_expr_stmt_or_stmt_list =
+                                parent.as_ref().map_or(false, |node| {
+                                    ast::ExprStmt::can_cast(node.kind())
+                                        | ast::StmtList::can_cast(node.kind())
```
Comment (davidsemakula) [inline crates/hir-ty/src/diagnostics/expr.rs:326]: @Veykril it appears if expr that don't end with a semi-colon aren't wrapped in an expr stmt, so I have to check for stmt list parent as well ... is the missing expr stmt a bug?

Comment (Veykril) [reply crates/hir-ty/src/diagnostics/expr.rs:326]: No that is correct, the block tail expression is not considered a statement by our grammar

File: crates/ide-diagnostics/src/handlers/remove_unnecessary_else.rs (hunk 1 above)
Comment (Veykril) [reply crates/ide-diagnostics/src/handlers/remove_unnecessary_else.rs:501]: That should be fine, but let's add a return type to `test3` so we don't have to supress the type mismatch diagnostic

**Comment by Veykril** [comment]
Thanks!
@bors r+

**Comment by Veykril** [comment]
meh need to rebase again I think
@bors r-

**Comment by Veykril** [comment]
Thanks!
@bors r+

**Pull request merged**